zerver/lib/stream_subscription.py

```diff
@@ -274,3 +275,60 @@ def subscriber_ids_with_stream_history_access(stream: Stream) -> Set[int]:
             stream.id, include_deactivated_users=False
         ).values_list("user_profile_id", flat=True)
     )
+
+
+def get_subscriptions_for_send_message(
+    *,
+    stream_id: int,
+    possible_wildcard_mention: bool,
+    possibly_mentioned_user_ids: AbstractSet[int],
+) -> QuerySet:
+    """
+    Note that this code optimizes an important use
+    case for large streams. Certain realms have
+    many long-term-idle users that are subscribed
+    to large streams (usually default streams), and
+    we don't acually have to send messages to most
+    of the users. It's expensive to process them
+    in a naive query, hence this function.
+
+    This returns a superset of the subscriptions that
+    will require action in the send-message codepath.
+
+    Basically, it excludes all long-term idle users
+    who don't have stream/push options set and who
+    also are not mentioned in the message.
+
+    For expediency, it also returns all users on the
+    stream who have ANY alert words, even if their
+    alert words aren't in the message. (We create
+    UserMessage rows for long-term-idle messages if
+    they have been alerted. Downstream logic will
+    filter out false alarms.)
+
+    We also get false alarms for certain messages
+    that **look** like they are mentioning a user
+    from a quick parse. Downstream logic will make
+    sure we filter out "falsely" mentioned users.
+    """
+
+    query = Subscription.objects.filter(
+        recipient__type=Recipient.STREAM,
+        recipient__type_id=stream_id,
+        active=True,
+        is_user_active=True,
+    )
```

Comment:
This can be `get_active_subscriptions_for_stream_id`, right?  It's the same query as a non-finalized QuerySet, and maybe better to avoid duplication?

Reply:
fixed